pubsub_logging/utils.py

```diff
@@ -0,0 +1,104 @@
+# limitations under the License.
+
+"""Utilities for the Python logging handlers."""
+
+import base64
+import os
+import sys
+import threading
+import time
+
+from googleapiclient import discovery
+from googleapiclient import errors
+
+import httplib2
+
+from oauth2client.client import GoogleCredentials
+
+from pubsub_logging.errors import RecoverableError
+
+
+PUBSUB_SCOPES = ["https://www.googleapis.com/auth/pubsub"]
+
+clients = threading.local()
+
+
+def compat_urlsafe_b64encode(v, force_py3_mode=False):
+    """A urlsafe ba64encode which is compatible with Python 2 and 3.
+
+    Args:
+      v: A string to encode.
+      force_py3_mode: If True, it will use the code for Python3.
+    Returns:
+      The encoded string.
+    """
+    if sys.version_info[0] == 3 or force_py3_mode:
+        return base64.urlsafe_b64encode(v.encode('UTF-8')).decode('ascii')
+    else:
+        return base64.urlsafe_b64encode(v)
+
+
+def get_pubsub_client(http=None):  # pragma: NO COVER
+    """Return a thread local Pub/Sub client.
+
+    Args:
+      http: httplib2.Http instance. Defaults to None.
+    Returns:
+      Cloud Pub/Sub client.
+    """
+    if not hasattr(clients, 'client'):
+        credentials = GoogleCredentials.get_application_default()
+        if credentials.create_scoped_required():
+            credentials = credentials.create_scoped(PUBSUB_SCOPES)
+        if not http:
+            http = httplib2.Http()
+        credentials.authorize(http=http)
+        clients.client = discovery.build('pubsub', 'v1beta2', http=http)
+    return clients.client
+
+
+def publish_body(client, body, topic, retry):
+    """Publishes the specified body to Cloud Pub/Sub.
+
+    Args:
+      client: Cloud Pub/Sub client.
+      body: Post body for Pub/Sub publish call.
+      topic: topic name that we publish the records to.
+      retry: number of retry upon intermittent failures.
+
+    Raises:
+      errors.HttpError When the Cloud Pub/Sub API call fails with
+                       unrecoverable reasons.
+      RecoverableError When the Cloud Pub/Sub API call fails with
+                       intermittent errors.
+    """
+    try:
+        before = time.time()
+        client.projects().topics().publish(
+            topic=topic, body=body).execute(num_retries=retry)
+        if os.environ.get('PSHANDLER_DEBUG', None):  # pragma: NO COVER
+            sys.stderr.write('Took %f secs for sending %d messages.\n' %
+                             (time.time() - before, len(body['messages'])))
+    except errors.HttpError as e:
+        if e.resp.status >= 400 and e.resp.status < 500:
+            # Publishing failed for some non-recoverable reason. For
+            # example, perhaps the service account doesn't have a
+            # permission to publish to the specified topic, or the topic
+            # simply doesn't exist.
+            raise
+        else:
+            # Treat this as a recoverable error.
+            raise RecoverableError()
```

Comment:
if retries are exceeded - the final error raised might likely be a 500 error, so a retry exceeded failure will pass silently here